insights/parsers/lssap.py

```diff
@@ -0,0 +1,118 @@
+
+"""
+lssap - Command
+============
+
+This module provides processing for the output of the ``lssap`` command on SAP systems.
+The spec handled by this command inlude::
+
+    "lssap"                     : CommandSpec("/usr/sap/hostctrl/exe/lssap")
+
+Class ``Lssap`` parses the output of the ``lssap`` command.  Sample
+output of this command looks like::
+
+ - lssap version 1.0 -
+ ==========================================
+   SID   Nr   Instance    SAPLOCALHOST                        Version                 DIR_EXECUTABLE
+   HA2|  16|       D16|         lu0417|749, patch 10, changelist 1698137|          /usr/sap/HA2/D16/exe
+   HA2|  22|       D22|         lu0417|749, patch 10, changelist 1698137|          /usr/sap/HA2/D22/exe
+   HA2|  50|       D50|         lu0417|749, patch 10, changelist 1698137|          /usr/sap/HA2/D50/exe
+   HA2|  51|       D51|         lu0417|749, patch 10, changelist 1698137|          /usr/sap/HA2/D51/exe
+
+Examples:
+    >>> lssap = shared[Lssap]
+    >>> lssap.instances()
+    ['D16', 'D22', 'D50', 'D51']
+    >>> ps_info.version('D51')
+    '749, path 10, changelist 1698137'
+    >>> lssap.is_instance()
+    True
+    >>> lssap.is_hana()
+    False
+    >>> lssap.data[3]
+    {'SID': 'HA2', 'Nr': '51', 'Instance': 'D51', 'SAPLOCALHOST': 'lu0417',
+     'Version': '749, path 10, changelist 1698137', 'DIR_EXECUTABLE': '/usr/sap/HA2/D51/exe'}
+    >>> lssap.data[3]['Instance']
+    'D51'
+"""
+from .. import Parser, parser, parse_table
+from insights.parsers import ParseException
+import re
+
+
+class SAPList(Parser):
+    """Base class implementing shared code."""
+
+    @property
+    def instances(self):
+        """list: Returns a list of instances from the Instance column."""
+        return [row["Instance"] for row in self.data if "Instance" in row]
+
+    @property
+    def sid(self):
+        """list: Returns a list of the SIDs from the SID column."""
+        return list(set(row["SID"] for row in self.data if "SID" in row))
+
+    def instance_types(self):
+        types = []
+        instances = "".join(self.instances)
+        netweaver = re.compile(r"D\d+")
+        hana = re.compile(r"HDB\d+")
+        ascs = re.compile(r"ASCS\d+")
+        if re.search(netweaver, instances):
+            types.append('netweaver')
+        if re.search(hana, instances):
+            types.append('hana')
+        if re.search(ascs, instances):
+            types.append('ascs')
+        else:
+            ParseException("No valid instance types found")
```

Comment:
Is it possible to determine if no valid instance types exist in `parse_content`, if so it would be better to raise the exception there, assuming no other useful data is available in that situation.  If all other data is valid then you need to document this exception so that a rule developer will understand that it will have to be caught in the rule.  We typically try to avoid this.

Reply:
Also, it doesn't look like you are testing this method directly.

Reply:
I agree.  I'd normally suggest computing this kind of stuff in parse_content, but you seem to do that in the LSSap parser class rather than in the SAPList class.  If it was all in one thing I'd do:
```python
    def parse_content(self, content):
        # remove lssap version and bar text from content
        clean_content = content[2:]  # There's a better way of doing this coming in parse_delimited_table...
        if len(clean_content) > 0 and "SID" in clean_content[0]:
            self.data = parse_table(clean_content, delim='|', header_split=False)
        else:
            raise ParseException("Lssap: Unable to parse {0} line(s) of content: ({1})".format(len(content), content))

        # Set attributes
        self.instances = [row["Instance"] for row in self.data if "Instance" in row]
        self.sids = list(set(row["SID"] for row in self.data if "SID" in row))
        self.instance_types = []
        if any(i.startswith('HDB') and i[3:].isdigit() for i in self.instances):
            self.instance_types.append('hana')
        if any(i.startswith('ASCS') and i[4:].isdigit() for i in self.instances):
            self.instance_types.append('ascs')
        if any(i.startswith('D') and i[1:].isdigit() for i in self.instances):
            self.instance_types.append('netweaver')
        if self.instance_types == []:
            raise ParseException("No valid instance types found")
```

But you probably  know this content and how it's going to change better than I do.